pkg/controller/csi_extension_server.go

```diff
@@ -0,0 +1,128 @@
+package controller
+
+import (
+	"context"
+	"errors"
+	"strings"
+	"time"
+
+	vgsext "github.com/dell/dell-csi-extensions/volumeGroupSnapshot"
+	"github.com/dell/gopowerstore"
+	log "github.com/sirupsen/logrus"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+const StateReady = "Ready"
+
+func (s *Service) CreateVolumeGroupSnapshot(ctx context.Context, request *vgsext.CreateVolumeGroupSnapshotRequest) (*vgsext.CreateVolumeGroupSnapshotResponse, error) {
+	log.Infof("CreateVolumeGroupSnapshot called with req: %v", request)
+
+	err := validateCreateVGSreq(request)
+	if err != nil {
+		log.Errorf("Error from CreateVolumeGroupSnapshot: %v ", err)
+		return nil, err
+	}
+	var reqParams gopowerstore.VolumeGroupSnapshotCreate
+	reqParams.Name = request.GetName()
+	reqParams.Description = request.GetDescription()
+	arr := strings.Split(request.SourceVolumeIDs[0], "/")[1]
+
+	var sourceVols []string
+	var volGroup gopowerstore.VolumeGroup
+	var snapsList []*vgsext.Snapshot
+
+	for _, v := range request.GetSourceVolumeIDs() {
+		sourceVols = append(sourceVols, strings.Split(v, "/")[0])
+	}
+	// To create volume group
+	vgParams := gopowerstore.VolumeGroupCreate{
+		Name:        request.GetName(),
+		Description: request.GetDescription(),
+		VolumeIds:   sourceVols,
+	}
+	resp, err := s.Arrays()[arr].GetClient().CreateVolumeGroup(ctx, &vgParams)
+	if err != nil {
+		if apiError, ok := err.(gopowerstore.APIError); !(ok && apiError.VolumeNameIsAlreadyUse()) {
+			return nil, status.Errorf(codes.Internal, "Error creating volume group: %s", err.Error())
+		}
+	}
+	if resp.ID != "" {
+		resp, err = s.Arrays()[arr].GetClient().CreateVolumeGroupSnapshot(ctx, resp.ID, &reqParams)
+		if err != nil {
+			if apiError, ok := err.(gopowerstore.APIError); !(ok && apiError.VolumeNameIsAlreadyUse()) {
+				return nil, status.Errorf(codes.Internal, "Error creating volume group snapshot: %s", err.Error())
+			}
+		}
+
+		volGroup, err = s.Arrays()[arr].GetClient().GetVolumeGroup(ctx, resp.ID)
+		log.Infof("%v", volGroup)
+		if err != nil {
+			if apiError, ok := err.(gopowerstore.APIError); !(ok && apiError.VolumeNameIsAlreadyUse()) {
+				return nil, status.Errorf(codes.Internal, "Error getting volume group snapshot: %s", err.Error())
+			}
+		}
+
+		for _, v := range volGroup.Volumes {
+			var snapState bool
+			if v.State == StateReady {
+				snapState = true
+			}
+			etime, _ := time.Parse(time.RFC3339, v.CreationTimeStamp)
+			int64Time := etime.Unix()
+			snapsList = append(snapsList, &vgsext.Snapshot{
+				Name:          v.Name,
+				SnapId:        v.ID + "/" + arr + "/" + strings.Split(request.SourceVolumeIDs[0], "/")[2],
```

Comment:
check for number of elements returned by Split() before accessing those elements.
same comment for line 78 also.

Reply:
done